src/charm.py

```diff
@@ -68,6 +68,8 @@ def CinderThreeParContext(charm_config, service):
         ctxt.append((
             'volume_driver',
             'cinder.volume.drivers.hpe.hpe_3par_iscsi.HPE3PARISCSIDriver'))
+    else:
+        return BlockedStatus('Invalid config (driver-type)')
```

Comment:
I agree it's a bit unusual, but I think it's OK to do something like this, since the return value can then be set as the status of the unit. If it weren't so, we'd need to catch a custom exception, then extract the message from it and then wrap a `BlockedStatus` around it, which is a bit more verbose and not worth it, IMO.

Reply:
Actually, thinking about this more, I'm against this design. If you want we can get a third opinion. (@ajkavanagh maybe?) 

What you describe is OK for now because the only caller of this function is interested in setting a `BlockedStatus`, so you spare code. What if tomorrow, there is another caller who wants to handle the error differently? They will have to do "if return value is of type BlockedStatus, then extract the message from it, parse it, and handle things accordingly". I understand why this is tempting now but I think this is how spaghetti emerges over time.

This design is suboptimal because it doesn't respect separation of concerns. This function should be concerned by things like unit status. This function shouldn't know what a unit status is. Its concern is configuration and context. There may be future callers who also aren't concerned by unit status and don't even "know" what this is.

Reply:
I'm in the @AurelienLourot camp here; this function is doing two things:

1. Validating the config.
2. Creating a context (for rendering, I suspect).

Therefore, it is mixing concerns.  Really, there should be two functions; one to parse and validate the config is valid, and then other to create the context (with programming asserts to cope with programming errors).

In this case, this function would replace the return BlockedStatus() with an assert to crash the charm (indicate a programming error).

A different function would validate the data.

So the structure of the code is flawed as it makes it harder to then use this function as you have to deal with the error case.  If you really want to validation and creating the context (which I don't recommend), then it should raise an Exception; really, no progress can be made if the config isn't valid for this function and it should break out using an exception.

Reply:
All very good points. I'll refactor this part to use exceptions.